### PR TITLE
feature: Use prettier api

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Relates SAP/cloud-sdk-backlog#736.

Use the prettier API to format .ts and other files. Source files like `.js` are excluded because this could break the source maps. In the past we created the map files before formating `.ts` which was wrong. Now it is correct.

Hence there are many changes in the generated clients files.

### CODE REVIEW

File: packages/generator-common/src/compiler.ts [hunk 1]
```diff
@@ -50,8 +53,39 @@ export async function transpileDirectory(
...
+    fileName,
+    text,
+    writeByteOrderMark,
+    onError?,
+    sourceFiles?,
+    data?
+  ) => {
+    const parsed = parse(fileName);
+    const promise = mkdir(parsed.dir, { recursive: true }).then(() => {
+      // All not emitted files like .md or .json should be already prettified.
```
Comment (marikaner) [inline packages/generator-common/src/compiler.ts:69]: [nth] General: In the section above we mention prettying, linting and formatting - I would prefer to reuse one word.

File: packages/generator-common/src/compiler.ts [hunk 2]
```diff
@@ -50,8 +53,39 @@ export async function transpileDirectory(
...
+      // Formatting emitted .js file could break source map -> skip these.
+      // Formatting the .map files also is not necessary increases only file size - only .d.ts files are nicer if linted
+      const usePrettier = getFileExtension(fileName) === 'd.ts' ? true : false;
+
+      // Prettier config should be cached from previous steps already.
+      // Emit also overwrites and copyright is either already there.
```
Comment (marikaner) [inline packages/generator-common/src/compiler.ts:75]: [req] I do not understand this sentence ;) Can you rephrase it?

File: packages/generator-common/src/file-writer/create-file.ts [hunk 3]
```diff
@@ -1,9 +1,138 @@
...
+
+/**
+ * This method considers also double dots like `.map.js`
+ * @param fileName
+ * @returns The complete file extension containing multiple dots
+ * @internal
+ */
+export function getFileExtension(fileName: string): string {
+  const fileType = parse(fileName).base.split('.').slice(1).join('.');
+  return fileType;
```
Comment (marikaner) [inline packages/generator-common/src/file-writer/create-file.ts:104]: [pp] Maybe return directly?
```suggestion
  return parse(fileName).base.split('.').slice(1).join('.');
```

File: packages/generator-common/src/file-writer/create-file.ts [hunk 4]
```diff
@@ -1,9 +1,138 @@
...
+}
+
+async function formatWithPrettier(
+  fileName: string,
+  content: string,
+  prettierConfigPath: string | undefined
+) {
+  const prettierConfig = await readPrettierConfig(prettierConfigPath);
```
Comment (marikaner) [inline packages/generator-common/src/file-writer/create-file.ts:112]: [q] We read the config for every file (I know, that we have a cache). But wouldn't it be simpler to just read it once and add it to the options? That would also make this function synchronous.

File: packages/generator-common/src/file-writer/create-file.ts [hunk 5]
```diff
@@ -1,9 +1,138 @@
...
+
+function addCopyrightHeader(content: string, withCopyright: boolean): string {
+  if (!withCopyright) {
+    return content;
+  }
+  return (
+    codeBlock`
+${getCopyrightHeader()}
+${content}
+` + unixEOL
```
Comment (marikaner) [inline packages/generator-common/src/file-writer/create-file.ts:133]: [req] I would put the `unixEOL` into the codeblock.
```suggestion
${content}
${unixEOL}
`
```
or:

```suggestion
${content}\n
`
```
[pp] After using it for a while, I see that I am not a fan of the `unixEOL`.  `\n` is just shorter, does not require escaping in strings and you don't need to import it. Maybe we could reconsider this.

File: packages/generator/package.json [hunk 6]
```diff
@@ -50,6 +50,7 @@
     "fast-xml-parser": "^4.0.11",
     "fs-extra": "^10.0.0",
     "glob": "^8.0.1",
+    "prettier": "^2.7.1",
```
Comment (marikaner) [inline packages/generator/package.json:53]: [req] This seems unnecessary? I didn't find any use in the generator package, only generator-common.

File: packages/generator-common/src/file-writer/create-file.ts [hunk 7]
```diff
@@ -1,9 +1,138 @@
...
+  /**
+   * Flag to indicate if copy write header is added.
+   */
+  withCopyright?: boolean;
+  /**
+   * Path to the prettier config with respect to the process.cwd().
+   */
+  prettierConfigPath?: string;
+  /**
+   * Flag to indicate if the file is prettified before.
```
Comment (marikaner) [inline packages/generator-common/src/file-writer/create-file.ts:36]: [req] Maybe we should avoid "prettier" as a verb and instead say "formatted with prettier".

**Review by marikaner — APPROVED** [review]
I like the new create options and I don't see any major blockers.
I wasn't sure if I got it right: When do we run the prettier? before transpilation or after? or both?

**Comment by FrankEssenberger** [comment]
> I like the new create options and I don't see any major blockers. I wasn't sure if I got it right: When do we run the prettier? before transpilation or after? or both?

Now we run it before the `.ts` files are written so before the transpilation to `.js`. Which is correct. Before we did it after the transpilation simply on all files which would break the source maps.

File: packages/generator-common/src/file-writer/create-file.ts (hunk 4 above)
Comment (FrankEssenberger) [reply packages/generator-common/src/file-writer/create-file.ts:111]: I also considered this. Since we have the options object now, it should not hurt I thing. Let me try that.

File: packages/generator/package.json (hunk 6 above)
Comment (FrankEssenberger) [reply packages/generator/package.json:53]: ups, yes

**Comment by marikaner** [comment]
> > I like the new create options and I don't see any major blockers. I wasn't sure if I got it right: When do we run the prettier? before transpilation or after? or both?
> 
> Now we run it before the `.ts` files are written so before the transpilation to `.js`. Which is correct. Before we did it after the transpilation simply on all files which would break the source maps.

But we also run it on the .d.ts files, so we run it twice?

File: packages/generator-common/src/compiler.ts (hunk 2 above)
Comment (FrankEssenberger) [reply packages/generator-common/src/compiler.ts:75]: removes because nonsense.
